waspc/src/Wasp/GeneratorMonad.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Wasp.GeneratorMonad
+  ( Generator,
+    GeneratorError (..),
+    GeneratorWarning (..),
+    GeneratorState (..),
+    catchGeneratorError,
+    logGeneratorError,
+    logAndThrowGeneratorError,
+    logGeneratorWarning,
+    runGenerator,
+  )
+where
+
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import qualified Control.Monad.Except as MonadExcept
+import Control.Monad.Identity (Identity (runIdentity))
+import Control.Monad.State (MonadState, StateT (runStateT), modify)
+
+-- | Generator is a monad transformer stack where we abstract away the underlying
+-- concrete monad transformers with the helper functions below. This will allow us
+-- to refactor and add more transformers (or swap them) without any caller changes.
+--
+-- The outer Either layer represents the fatal error that halted generation.
+-- There may optionally be other errors or warnings logged in the State.
+newtype Generator a = Generator
+  { _runGenerator :: ExceptT GeneratorError (StateT GeneratorState Identity) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState GeneratorState,
+      MonadError GeneratorError
+    )
+
+data GeneratorState = GeneratorState
+  { warnings :: [GeneratorWarning],
+    errors :: [GeneratorError]
+  }
+
+data GeneratorError
+  = GenericGeneratorError String
+  | PackageJsonGenerationError String
+  | PrismaGenerationError String
+
+instance Show GeneratorError where
+  show (GenericGeneratorError e) = e
+  show (PackageJsonGenerationError e) = "package.json Generation Error: " ++ e
+  show (PrismaGenerationError e) = "Prisma Generation Error: " ++ e
+
+data GeneratorWarning = GenericGeneratorWarning String
```

Comment:
Hm, good question. I am not sure on how best to handle info atm. Maybe this is a "cross that bridge when we need it" situation? But if we really do, and it's painful, maybe we should consider it now... hmm.

The other option I can see is to do something more flexible to start (but potentially not using/complicating things a bit) and combine them all into one thing like:

```haskell
data LogLevel = Error | Warning | Info

data GeneratorLog = GenericGeneratorLog LogLevel String
```

It's not terrible... 🤔 But complicates pattern matching and does make it less obvious at the type level what is going on. I'll sleep on it, curious about your thoughts.

Reply:
Yeah, I think this would be alternative way to go. As you said, it forces to check if there is any "log" with error severity, instead of just pattern matching to see if error array is empty. Not a huge deal probably but still.

I would certainly not push in this direction right now, I think it is too early.
Great thing is that we have these functions for throwing warnings and errors that can stay as they are even if we modify this part of how are errors organized and returned! So that is great.

TLDR; ok let's not worry about this and leave it for later, I think we are in good place to easily modify this in the future as needed.

Reply:
Sounds good, and agreed - thanks! 👍🏻